[LBSE] Enable accelerated transform animations for SVG renderers

nikolaszimmermann opened the pull request:
#### bde602f47049d0712b479d925ea19e0962e05f7f
<pre>
[LBSE] Enable accelerated transform animations for SVG renderers
<a href="https://bugs.webkit.org/show_bug.cgi?id=258122">https://bugs.webkit.org/show_bug.cgi?id=258122</a>

Reviewed by Rob Buis.

Accelerated transform animations were not active for SVG renderers
in LBSE. RenderLayerBacking::startAnimation() only used accelerated
mode for RenderBoxModelObject derived renderers -- extend that to
all SVG layer awarerenderers. Forgot to upstream this half a year
ago :-(

Also make sure that &apos;transform-box&apos; is respected, when calculating
the bounding box, against which e.g. &apos;transform-origin&apos; percentage
values are resolved. This needs more testing to verify this outside
of SVG content in future -- need to coordinate with WPT.

Covered by existing tests, performance difference is significant.

Canonical link: <a href="https://commits.webkit.org/268111@main">https://commits.webkit.org/268111@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6cb774a237c43f2eacbc466c3b9f73346a2c5ae3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19706 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20626 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17560 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19239 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19124 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21508 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16346 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17096 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17373 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17268 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21445 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17867 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16930 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4449 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21296 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23377 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17706 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -4025,16 +4027,20 @@ bool RenderLayerBacking::startAnimation(double timeOffset, const Animation& anim
     bool didAnimate = false;
 
-    if (hasRotate && m_graphicsLayer->addAnimation(rotateVector, snappedIntRect(m_owningLayer.rendererBorderBoxRect()).size(), &animation, keyframes.animationName(), timeOffset))
+    auto referenceBoxRect = renderer().transformReferenceBoxRect(renderer().style());
```
Replacing rendererBorderBoxRect with transformReferenceBoxRect seems like a big step. Can you explain if anything changes for non SVG? Basically I need a bit more explanation about the change in behaviour.

nikolaszimmermann replied:
Good question, need to explain a bit, why it is not a behavior change per-se, but only fixes a bug: 'transform-box' is ignored, where it shouldn't. 

transformReferenceBoxRect() calls referenceBoxRect() taking into account 'transform-box' -- which is missing at present --don't know if it has WPT coverage (transform-box has in general, however not sure for that particular CSS animation codepath). I think I created test cases in LBSE downstream for that, need to check it again.

**RenderElementInlines.h:**
```
inline FloatRect RenderElement::transformReferenceBoxRect(const RenderStyle& style) const { return referenceBoxRect(transformBoxToCSSBoxType(style.transformBox())); }
```

**RenderBox.cpp:**

```
FloatRect RenderBox::referenceBoxRect(CSSBoxType boxType) const
{
    switch (boxType) {
    case CSSBoxType::ContentBox:
    case CSSBoxType::FillBox:
        return contentBoxRect();
    case CSSBoxType::PaddingBox:
        return paddingBoxRect();
    case CSSBoxType::MarginBox:
        return marginBoxRect();
    // stroke-box, view-box compute to border-box for HTML elements.
    case CSSBoxType::StrokeBox:
    case CSSBoxType::ViewBox:
    case CSSBoxType::BorderBox:
    case CSSBoxType::BoxMissing:
        return borderBoxRect();
    }

    ASSERT_NOT_REACHED();
    return { };
}
```

rwlbuis reviewed (approved):
Looks good, but I think the commit message needs improvement.
First, you should mention the "transform-box" bug fix.

>Covered by existing tests, performance difference is striking.

Difference to what? Also "striking" is a bit subjective, something like "significant" may be better.

nikolaszimmermann committed:
[LBSE] Enable accelerated transform animations for SVG renderers
https://bugs.webkit.org/show_bug.cgi?id=258122

Reviewed by Rob Buis.

Accelerated transform animations were not active for SVG renderers
in LBSE. RenderLayerBacking::startAnimation() only used accelerated
mode for RenderBoxModelObject derived renderers -- extend that to
all SVG layer awarerenderers. Forgot to upstream this half a year
ago :-(

Also make sure that 'transform-box' is respected, when calculating
the bounding box, against which e.g. 'transform-origin' percentage
values are resolved. This needs more testing to verify this outside
of SVG content in future -- need to coordinate with WPT.

Covered by existing tests, performance difference is significant.

Canonical link: https://commits.webkit.org/268111@main
